File: 백준/Gold/2533. 사회망 서비스（SNS）/사회망 서비스（SNS）.py
import sys

try:
    sys.stdin = open('boj.in', encoding='utf-8')
except FileNotFoundError:
    pass

input = lambda: sys.stdin.readline().rstrip()
sys.setrecursionlimit(10 ** 6 + 1000)

class Graph:
    def __init__(self, n):
        self.n = n
        self.children = [[] for _ in range(n + 1)]
        self.edges = [set() for _ in range(n + 1)]

    def add_edge(self, a, b):
        self.edges[a].add(b)
        self.edges[b].add(a)

    def dfs(self, v):
        stack = [v]
        visited = {v}
        while stack:
            v = stack.pop()
            for nxt in self.edges[v]:
                if nxt in visited: continue
                stack.append(nxt)
                visited.add(nxt)
                self.children[v].append(nxt)
                

def solve(g, u, dp):
    if dp[u][0] == -1:
        dp[u][0] = 0
        for c in g.children[u]:
            dp[u][0] += solve(g, c, dp)[1]

    if dp[u][1] == -1:
        dp[u][1] = 1
        for c in g.children[u]:
            dp[u][1] += min(solve(g, c, dp))

    return dp[u][0], dp[u][1]

n = int(input())
g = Graph(n)
dp = [[-1, -1] for _ in range(n + 1)]

for _ in range(n - 1):
    p, c = map(int, input().split())
    g.add_edge(p, c)

g.dfs(1)
print(min(solve(g, 1, dp)))

File: 백준/Gold/2533. 사회망 서비스（SNS）/사회망 서비스（SNS）.cc
#include <bits/stdc++.h>

#define endl '\n'
#define FASTIO ios::sync_with_stdio(0), cin.tie(0), cout.tie(0)

void _init() {
#ifndef BOJ
    freopen("boj.in", "r", stdin);
#endif // BOJ
}

using namespace std;

using ll = long long;
using vi = vector<int>;
using pii = pair<int, int>;

constexpr int INF = 2e8;

int _min(pii& p) {
    return (p.first < p.second) ? p.first : p.second;
}

class Graph {
public:
    int n;
    vector<vi> children;
    vector<set<int>> edges;
    
    Graph(int n) : n(n) {
        children = vector<vi>(n + 1);
        edges = vector<set<int>>(n + 1);
    }

    void add_edge(int u, int v) {
        edges[u].insert(v);
        edges[v].insert(u);
    }

    void dfs(int v) {
        vi st;
        set<int> visited;
        st.push_back(v);
        visited.insert(v);
        while (!st.empty()) {
            v = st.back();
            st.pop_back();
            for (int nxt : edges[v]) {
                if (visited.find(nxt) != visited.end()) continue;
                visited.insert(nxt);
                st.push_back(nxt);
                children[v].push_back(nxt);
            }
        }
    }
};

pii& solve(Graph& g, int u, vector<pii>& dp) {
    if (dp[u].first == -1) {
        dp[u].first = 0;
        for (int c : g.children[u])
            dp[u].first += solve(g, c, dp).second;
    }

    if (dp[u].second == -1) {
        dp[u].second = 1;
        for (int c : g.children[u])
            dp[u].second += _min(solve(g, c, dp));
    }

    return dp[u];
}

int main() {
    FASTIO;
    _init();

    int n; cin >> n;
    Graph g(n);
    vector<pii> dp(n + 1, pii(-1, -1));

    while (n --> 1) {
        int u, v; cin >> u >> v;
        g.add_edge(u, v);
    }

    g.dfs(1);
    cout << _min(solve(g, 1, dp)) << endl;

    return 0;
}
